File: src/client/java/net/bewis09/bewisclient/mixin/client/MouseMixin.java
package net.bewis09.bewisclient.mixin.client;

import net.bewis09.bewisclient.impl.functionalities.zoom.Zoom;
import net.bewis09.bewisclient.impl.widget.CPSWidget;
import net.minecraft.class_11910;
import net.minecraft.client.Mouse;
import net.minecraft.util.math.MathHelper;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

@Mixin(Mouse.class)
public class MouseMixin {
    @Inject(method = "onMouseButton", at = @At("HEAD"))
    private void bewisclient$onMouseButton(long window, class_11910 arg, int action, CallbackInfo ci) {
        if (action != 1) return;

        if (arg.button() == 0) CPSWidget.INSTANCE.getLeftMouseList().add(System.currentTimeMillis());
        if (arg.button() == 1) CPSWidget.INSTANCE.getRightMouseList().add(System.currentTimeMillis());
    }

    @Inject(method = "onMouseScroll", at = @At("HEAD"), cancellable = true)
    private void bewisclient$onMouseScroll(long window, double horizontal, double vertical, CallbackInfo ci) {
        if (Zoom.INSTANCE.isUsed()) {
            Zoom.INSTANCE.getFactorAnimation().set("factor", MathHelper.clamp(Zoom.INSTANCE.getFactorAnimation().getWithoutInterpolation("factor") - (float) vertical * 0.02f, .009f, .4f));

            ci.cancel();
        }
    }
}

File: src/client/kotlin/net/bewis09/bewisclient/mixin/client/MouseMixin.kt
// @VersionReplacement

package net.bewis09.bewisclient.mixin.client

import com.mojang.blaze3d.platform.InputConstants
import net.bewis09.bewisclient.features.utilities.Zoom.factorAnimation
import net.bewis09.bewisclient.features.utilities.Zoom.isUsed
import net.bewis09.bewisclient.widget.impl.CPSWidget.leftMouseList
import net.bewis09.bewisclient.widget.impl.CPSWidget.rightMouseList
import net.minecraft.client.MouseHandler
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo

@Mixin(MouseHandler::class)
class MouseMixin {
    // @[1.21.8] "onPress" @[] "onButton"
    @Inject(method = [/*[@]*/"onButton"/*[!@]*/], at = [At("HEAD")])
    // @[1.21.8] button: Int, action: Int, modifiers: Int @[] rawButtonInfo: net.minecraft.client.input.MouseButtonInfo, action: Int
    private fun bewisclientOnMouseButton(handle: Long, /*[@]*/rawButtonInfo: net.minecraft.client.input.MouseButtonInfo, action: Int/*[!@]*/, ci: CallbackInfo?) {
        if (action != 1) return

        // @[1.21.8] button @[] rawButtonInfo.button()
        when (/*[@]*/rawButtonInfo.button()/*[!@]*/) {
            InputConstants.MOUSE_BUTTON_LEFT -> leftMouseList.add(System.currentTimeMillis())
            InputConstants.MOUSE_BUTTON_RIGHT -> rightMouseList.add(System.currentTimeMillis())
        }
    }

    @Inject(method = ["onScroll"], at = [At("HEAD")], cancellable = true)
    private fun bewisclientOnMouseScroll(handle: Long, horizontal: Double, vertical: Double, ci: CallbackInfo) {
        if (isUsed()) {
            factorAnimation.set((factorAnimation.getWithoutInterpolation() - vertical.toFloat() * 0.02f).coerceIn(.009f, .4f))

            ci.cancel()
        }
    }
}